### PR TITLE
feat(shiny create): Support simpler syntax for `--github` flag

### PR DESCRIPTION
**Pull request by gadenbuie**
This PR replaces the requirement for a full URL to a template hosted in a GitHub repo, replacing it with a shorter, more familiar syntax.

```bash
# In addition to the current incantation
shiny create -g https://github.com/posit-dev/py-shiny-templates/tree/main/map-distance

# These are now valid
shiny create --github posit-dev/py-shiny-templates/map-distance

shiny create --github posit-dev/py-shiny-templates/map-distance@main
```

Because there are many common ways to specify the `ref` and `path` portions of the GitHub spec, I ended up adding support for more than a few variations

* `{repo_owner}/{repo_name}@{ref}:{path}`
* `{repo_owner}/{repo_name}:{path}@{ref}`
* `{repo_owner}/{repo_name}:{path}`
* `{repo_owner}/{repo_name}/{path}`
* `{repo_owner}/{repo_name}/{path}@{ref}`
* `{repo_owner}/{repo_name}/{path}?ref={ref}`

Alternatively, the `--template` flag can now be combined with the `--github` flag, so these iterations are now the same as the above:

```bash
shiny create --github posit-dev/py-shiny-templates --template map-distance

shiny create --github posit-dev/py-shiny-templates@main --template map-distance
```

I'd really like to support listing available templates and asking the user to pick one, so that we could support

```bash
shiny create --github posit-dev/py-shiny-templates
```

with a missing `--template` flag. `shiny create` would download `posit-dev/py-shiny-templates`, find the available templates in the repo, and provide a menu to users.

While here, I've also improved the use of formatting and color in the CLI interface. Here's an example of the output seen when creating the `map-distance` template:

image

### CODE REVIEW

File: shiny/_main.py [hunk 1]
```diff
@@ -602,21 +605,28 @@ def create(
...
     github: Optional[str] = None,
-    dir: Optional[str | Path] = None,
+    dir: Optional[Path] = None,
     package_name: Optional[str] = None,
 ) -> None:
     from ._template_utils import template_query, use_git_template
 
-    if github is not None and template is not None:
-        raise click.UsageError("You cannot provide both --github and --template")
+    valid_templates = {**app_template_choices, **package_template_choices}.values()
```
Comment (cpsievert) [inline shiny/_main.py:613]: Move this down to 622

File: shiny/_main.py [hunk 2]
```diff
@@ -602,21 +605,28 @@ def create(
...
 
     if isinstance(dir, str):
         dir = Path(dir)
 
     if github is not None:
-        use_git_template(github, mode, dir)
+        use_git_template(github, template=template, mode=mode, dest_dir=dir)
         return
```
Comment (cpsievert) [inline shiny/_main.py:620]: I think I'd prefer an if else here

File: shiny/_template_utils.py [hunk 3]
```diff
@@ -104,34 +103,173 @@ def download_and_extract_zip(url: str, temp_dir: Path):
...
+    # We have exactly one directory and no other files
+    if len(directories) == 1 and len(files) == 0:
+        return directories[0]
+
+    return temp_dir
+
 
 def use_git_template(
-    url: str, mode: Optional[str] = None, dest_dir: Optional[Path] = None
+    url: str,
```
Comment (cpsievert) [inline shiny/_template_utils.py:122]: Could we rename this to `github` or something?

File: shiny/_template_utils.py [hunk 4]
```diff
@@ -104,34 +103,173 @@ def download_and_extract_zip(url: str, temp_dir: Path):
...
+            )
+
+        template_dir = template_dir / spec.path
 
-        template_dir = os.path.join(
-            temp_dir, f"{repo_name}-{branch_name}", subdirectory
+        if not os.path.exists(template_dir):
+            raise click.ClickException(
+                f"Template directory '{cli_input(spec.path)}' does not exist "
+                + f"in the {cli_field(spec.repo_owner + '/' + spec.repo_name)} repository."
```
Comment (cpsievert) [inline shiny/_template_utils.py:167]:
```suggestion
                + f"in the {cli_field(spec_cli)} repository."
```

File: shiny/_template_utils.py [hunk 5]
```diff
@@ -104,34 +103,173 @@ def download_and_extract_zip(url: str, temp_dir: Path):
...
+        # refs/heads/{branch}.zip, refs/tags/{tag}.zip, or just {commit}.zip.
+        # It's hard to tell in advance which one is intended, so we try all three in
+        # succession, using the first that works.
+        for zip_url in github_zip_url(spec):
+            try:
+                template_dir = download_and_extract_zip(zip_url, temp_dir)
+                success = True
+                break
+            except Exception:
+                pass
```
Comment (cpsievert) [inline shiny/_template_utils.py:154]: Could accumulate execptions and include them if not success?

Comment (gadenbuie) [reply shiny/_template_utils.py:161]:
```
❯ shiny create --github baslkjsadfas/aslalskdasdf
ℹ Using GitHub repository baslkjsadfas/aslalskdasdf.
Error: Failed to download repository from GitHub baslkjsadfas/aslalskdasdf. Please check the URL or GitHub spec and try again.
We tried the following URLs
  https://github.com/baslkjsadfas/aslalskdasdf/archive/refs/heads/HEAD.zip:
    HTTP Error 404: Not Found for url: https://github.com/baslkjsadfas/aslalskdasdf/archive/refs/heads/HEAD.zip
  https://github.com/baslkjsadfas/aslalskdasdf/archive/refs/tags/HEAD.zip:
    HTTP Error 404: Not Found for url: https://github.com/baslkjsadfas/aslalskdasdf/archive/refs/tags/HEAD.zip
  https://github.com/baslkjsadfas/aslalskdasdf/archive/HEAD.zip:
    HTTP Error 404: Not Found for url: https://github.com/baslkjsadfas/aslalskdasdf/archive/HEAD.zip
```

File: shiny/_main.py [hunk 6]
```diff
@@ -602,22 +605,18 @@ def create(
     template: Optional[str] = None,
     mode: Optional[str] = None,
     github: Optional[str] = None,
-    dir: Optional[str | Path] = None,
+    dir: Optional[Path] = None,
```
Comment (cpsievert) [inline shiny/_main.py:608]: Are we sure this is always a `Path`? If so, why keep the `str` runtime check?

Comment (gadenbuie) [reply shiny/_main.py:608]: Oops, I didn't mean to remove `str` from the input typing. I do think we could use

```
type=click.Path(dir_okay=True, file_okay=False)
```

for the `--dir` option. This does guarantee that `dir` is then a `Path()`. I'd still be on the fence about allowing `str`, wdyt?

File: shiny/_template_utils.py [hunk 7]
```diff
@@ -104,34 +111,177 @@ def download_and_extract_zip(url: str, temp_dir: Path):
...
+            + cli_field("{repo_owner}/{repo_name}@{ref}:{path}")
+            + ".",
+            param_hint="--github",
+        )
+
+    repo_owner, _, repo_name = parts[:3]
+    path = ""
+    ref = ""
+
+    which = "path"
```
Comment (cpsievert) [inline shiny/_template_utils.py:237]: Seems like you could get rid of the `which`?

Comment (gadenbuie) [reply shiny/_template_utils.py:237]: That for loop has two parts: 

1. if we're looking at a separator, we decide what the separator means
2. if it's not a separator, decide where to add the part we're considering

So I can't get rid of `which="path"`, aka the initial assumption, but we could move it other places, but I'm not seeing a scenario that I find to be more clear. What do you have in mind?

File: shiny/_template_utils.py [hunk 8]
```diff
@@ -104,34 +111,177 @@ def download_and_extract_zip(url: str, temp_dir: Path):
...
+    * {repo_owner}/{repo_name}/{path}
+    * {repo_owner}/{repo_name}/{path}@{ref}
+    * {repo_owner}/{repo_name}/{path}?ref={ref}
+    """
+
+    # We split the spec into parts, using a capture group so that the splitting
+    # characters are retained in the path parts. Then we know that {repo_owner} /
+    # {repo_name} come first in the first three parts, reducing the problem to parsing
+    # ref and path in the remaining parts.
+    parts = re.split(r"(:|@|[?]ref=|/)", spec)
```
[the remark this answers is not in the document]
Comment (gadenbuie) [reply shiny/_template_utils.py:222]: Yeah and org/repo owner names have similar constraints

File: shiny/_template_utils.py (hunk 7 above)
Comment (cpsievert) [reply shiny/_template_utils.py:237]: Ohhh, duh, sorry for the noise.

File: shiny/_main.py (hunk 6 above)
Comment (cpsievert) [reply shiny/_main.py:608]: Ok, yea, that seems sensible, and I don't see any harm in keeping the `str` type.

File: shiny/_template_utils.py [hunk 9]
```diff
@@ -257,24 +444,21 @@ def directory_prompt(
...
 
 
-def build_path_string(*path: str):
+def path_rel_wd(*path: str):
     """
-    Build a path string that is valid for the current OS
+    Path relative to the working directory, formatted for the current OS
     """
-    return os.path.join(".", *path)
+    return os.path.join(".", *(path or [""]))
```
Comment (cpsievert) [inline shiny/_template_utils.py:461]: I don't quite follow why this change is here -- both versions seem to fail on `None` and work with `""`?

Comment (gadenbuie) [reply shiny/_template_utils.py:472]: I mean I'm not too sure either, this helper function seemed to exist to create `./` using OS-specific separators. Before my change, we were calling this function `build_path_string()` and we would call it like `build_path_string("")` in the base case. 

That seems to indicate the goal is to make `"./"` and not just `"."`. The output is used as the starting value in destination directory prompts, so this change makes sense.

As it stands in this PR, you either call `path_rel_wd()` without arguments or you give it strings of path parts that are joined together.

Comment (cpsievert) [reply shiny/_template_utils.py:472]: Ahh I got it, so this is really about calling the function with no args:

```python
>>> path_rel_wd()
'./'
>>> build_path_string()
'.'
```

**Review by cpsievert — APPROVED** [review]
LGTM once tests are passing
